owasp-top10-2017-apps/a2/saidajaula-monster/app/app.py

```diff
@@ -102,15 +117,38 @@ def login():
         if not password.validate_password(result[0]):
             return "Login failed! \n"
 
-        cookie_dic = {"permissao": result[1], "username": form_username}
-        cookie = json.dumps(cookie_dic)
-        hash_cookie = hashlib.sha256(cookie.encode('utf-8')).hexdigest()
-        cookie_done = '.'.join([cookie,hash_cookie])
-        cookie_done = base64.b64encode(str(cookie_done).encode("utf-8"))
+        access_token, refresh_token = create_tokens(form_username, result[1])
+
         resp = make_response("Logged in!")
-        resp.set_cookie("sessionId", cookie_done)
+        resp.set_cookie("access_token", access_token)
+        resp.set_cookie("refresh_token", refresh_token, path="/refresh-token")
         return resp
 
+@app.route("/refresh-token", methods=['GET'])
+def refresh_token():
+    cookie = request.cookies.get("refresh_token", "")
+    try:
+        payload = jwt.decode(cookie, SECRET_KEY, algorithms='HS256')
+        if len(payload) > 2:
+            return redirect("/login")
+
+        username = payload["username"]
+        result, success = database.get_user(username)
+        if not success:
+            return redirect("/login")
+
+        new_access_token, new_refresh_token = create_tokens(username, result[1])
+
+        referer = request.referrer
+        if referer != None:
+            resp = redirect(referer)
+        else:
+            resp = redirect("user")
+        resp.set_cookie("access_token", new_access_token)
+        resp.set_cookie("refresh_token", new_refresh_token, path="/refresh-token")
```

Comment:
You **should not** set a new refresh token here. 
In this route, the app only checked the current refresh token, and the refresh token should only be used to generate a new access token. In order to generate a new **refresh token**, the user would have to log in again

Reply:
@gustavocovas, I was in doubt because in the [oauth site](https://www.oauth.com/oauth2-servers/access-tokens/refreshing-access-tokens/) say that returne a new refresh token is optional.

```
The server may issue a new refresh token in the response,
but if the response does not include a new refresh token,
the client assumes the existing refresh token will still be valid.
```

Reply:
How I don't put the refresh token in a blacklist, perhaps it should not set a new refresh token.

Reply:
Sure. But our app is not implementing the OAuth flow with access and refresh tokens, but a much simpler protocol (which was not clearly defined, sorry for that 😕). I did not understand what you said about a blacklist.

In the OAuth protocol, the authorization server has other means to authenticate the client when refreshing an access token, as the [specification says](https://tools.ietf.org/html/rfc6749#section-6).

Here we implemented a much simpler protocol, where we wanted to have short-lived access tokens and long-lived refresh tokens, where the later should only be used to obtain new access tokens. We would have to do a threat modelling exercise to think about the implications of having each of these tokens leaked, or whether we should issue a new refresh token while issuing a new access token. 

In the context of the original vulnerability (which was already mitigated), I think we did a good job learning about the concept of access and refresh tokens, and how they are used in a production-ready standard such as OAuth2.